### PR TITLE
Feature/switch 911 write org creator service

### PR DESCRIPTION
**Pull request by ahmedolaibrahim**
- Initial implementation of consumer endpoint to  process Claims Creation Request event from Nats server
- Logger Implementation

### CODE REVIEW

File: src/logger/logger.service.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import {
+  createLogger,
+  format,
+  Logger as WinstonLogger,
+  transports,
+} from 'winston';
+
+import { ConfigService } from '@nestjs/config';
+
+@Injectable({ scope: Scope.TRANSIENT })
```
Comment (jrhender) [inline src/logger/logger.service.ts:16]: Do you know why the `TRANSIENT` scope? I see we've got this in the `iam-cache-server` but I'm not sure why this is 🤔  . I don't see any reason in principle why the Logger provide should not be shared across consumers

File: src/org-creator/orgCreator.controller.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  ) {
+    this.logger.setContext(OrgCreatorController.name);
+  }
+
+  @EventPattern('*.claim.exchange')
+  async createOrg(@Payload() message: ClaimRequestEventDto) {
+    const { token } = message;
+    const {
+      payload: { claimData },
+    } = jwt.decode(token, { complete: true });
```
Comment (jrhender) [inline src/org-creator/orgCreator.controller.ts]: Minor point but if you only need the payload, do you need `{ complete: true }`?

File: src/org-creator/orgCreator.controller.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    const permittedEventRole = this.configService.get<string>(
+      'PERMITTED_ORG_CREATOR_ROLE',
+    );
+    if (claimData?.claimType !== permittedEventRole) {
```
Comment (jrhender) [inline src/org-creator/orgCreator.controller.ts]: What does `configService` return if `PERMITTED_ORG_CREATOR_ROLE` is not set? Would it be possible if both `claimType` and `permitedEventRole` are empty then `claimData?.claimType !== permittedEventRole` would not be hit? Maybe worth adding a test for this?

**Review by jrhender — COMMENTED** [review]
Looks good so far!

File: src/logger/logger.service.ts (hunk 1 above)
Comment (ahmedolaibrahim) [reply src/logger/logger.service.ts:16]: I think there's probably an unwritten rule against allowing enhancers to be REQUEST scoped. I remember I had issues in the past with this and the service was returning undefined because i changed it to REQUEST scope but you're right there isn't any reason in principle why it shouldn't be shared across consumers.

File: src/org-creator/orgCreator.controller.ts (hunk 2 above)
Comment (ahmedolaibrahim) [reply src/org-creator/orgCreator.controller.ts]: you're right,  I used it at first when testing, should have removed it after.

File: src/org-creator/orgCreator.controller.ts (hunk 3 above)
Comment (ahmedolaibrahim) [reply src/org-creator/orgCreator.controller.ts]: hmmm, for the claim request creation event we'll probably always have `permittedEventRole` because the DTO validation would ensure that even if it's an empty string `' '`.  so even if  `PERMITTED_ORG_CREATOR_ROLE ` is null/undefined, `undefined !== ' '` would always be true.  It's a valid edge case to have though. I'll add it in the test.

Comment (JGiter) [reply src/org-creator/orgCreator.controller.ts]: Shouldn't we rely on cache-server for org creator role verification?

Comment (jrhender) [reply src/org-creator/orgCreator.controller.ts]: @JGiter I don't see how the cache-server helps here, to be honest. Could you elaborate on how you see the cache-server playing a role please?

Comment (JGiter) [reply src/org-creator/orgCreator.controller.ts]: Why not ask claims of requester and check if there is approved claim with PERMITTED_ORG_CREATOR?

Comment (ahmedolaibrahim) [reply src/org-creator/orgCreator.controller.ts]: @JGiter so at that point, the claims request would not have been approved so there probably won't be an approved claim to check for.

File: example.env [hunk 4]
```diff
@@ -1,5 +1,14 @@
 # NESTJS CONFIG
```
Comment (jrhender) [inline example.env:1]: Please add a `.env.dev` with sample values for dev. For `RPC_URL`, use "https://volta-rpc.com" and leave `PRIVATE_KEY` blank for now.

File: example.env [hunk 5]
```diff
@@ -1,5 +1,14 @@
...
 NESTJS_PORT=
```
Comment (jrhender) [inline example.env:2]: In my opinion, would be good to have documentation on what each of these properties does. I know we don't have in other apps but I think it is a good practice. What do you think?

File: src/org-creator/mock/mock-data.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      value: 'mctesty',
+    },
+  ],
+  claimType: 'test.role.testy.apps.testorg.iam.ewc',
+  claimTypeVersion: '1.0.0',
+};
+
+export const createClaimRequest = {
+  id: chance.guid({ version: 4 }),
+  token: chance.apple_token(),
```
Comment (jrhender) [inline src/org-creator/mock/mock-data.ts]: Why the use of `apple_token`? Just to get a random string or is there some other reason?

File: src/org-creator/orgCreator.controller.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+  ) {
+    this.logger.setContext(OrgCreatorController.name);
+  }
+
+  @EventPattern('*.claim.exchange')
+  async createOrg(@Payload() message: ClaimRequestEventDto): Promise<boolean> {
+    const requestObject = plainToClass(ClaimRequestEventDto, message);
+    const errors = await validate(requestObject, {
+      whitelist: true,
+      forbidNonWhitelisted: true,
```
Comment (jrhender) [inline src/org-creator/orgCreator.controller.ts:36]: I don't think you need both `whitelist` and `forbidNonWhiteListed`. `forbidNonWhiteListed` seems to override `whitelist`? https://docs.nestjs.com/techniques/validation

**Review by jrhender — COMMENTED** [review]
Looks good! Just a few comments/questions.

Also, would it be possible update the README for this repo when you have time please?

File: example.env (hunk 4 above)
Comment (ahmedolaibrahim) [reply example.env:1]: ok cool. will do!

File: src/org-creator/mock/mock-data.ts (hunk 6 above)
Comment (ahmedolaibrahim) [reply src/org-creator/mock/mock-data.ts]: yeah this is just to get a random string.

File: src/org-creator/orgCreator.controller.ts (hunk 7 above)
Comment (ahmedolaibrahim) [reply src/org-creator/orgCreator.controller.ts:36]: You actually need to. It's not stated in the nestjs docs but it is in the class-validator docs. here's the [link](https://www.npmjs.com/package/class-validator#whitelisting). I also tested it to confirm.

File: .env.dev [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# NESTJS CONFIG
+NESTJS_PORT=9000
+
+# NATS.IO CONFIG
+NATS_CLIENTS_URL=nats://identityevents-dev-nats.energyweb.org:4222
+
+REQUEST_NEW_ORG_ROLE=mytest.roles.testy.iam.ewc
+VOLTA_CHAIN_ID=73799
+CACHE_SERVER_URL=https://identitycache-dev.energyweb.org/
+RPC_URL=https://volta-rpc.com
```
Comment (jrhender) [inline .env.dev]: better to have it be actual RPC url
```suggestion
RPC_URL=https://volta-rpc.energyweb.org
```

**Review by jrhender — APPROVED** [review]
Looks good, nice work! Just need to deploy now 👍
